docs/design/rosbag2_snapshot_recorder/rosbag2_snapshot_recorder.md

```diff
@@ -0,0 +1,738 @@
+ros2 service call /rosbag2_recorder/snapshot/duration rosbag2_interfaces/srv/SetDuration "{duration: 50.0}"
+
+# Clear snapshot buffer
+ros2 service call /rosbag2_recorder/snapshot/clear_buffer rosbag2_interfaces/srv/ClearBuffer "{ }"
+
+# Pause messages buffering
+ros2 service call /rosbag2_recorder/snapshot/pause rosbag2_interfaces/srv/Pause "{ }"
+
+# Resume messages buffering
+ros2 service call /rosbag2_recorder/snapshot/resume rosbag2_interfaces/srv/Resume "{ }"
+```
+
+## Design
+
+### Main assumptions
+
+* Snapshot functionality makes use of producer-consumer cache patter which must be always turned on.
+* Main idea is to use `split_bag()` functionality in `SequentialWriter` for creating snapshots. In current implementation storage `open()` must be called first before any other function. This design manipulates start_time and duration metadata in `SnapshotWriter::write` just to match the state of buffered mesages.
+* Additionally this design forces that before `Recorder` node shut down, snapshot is getting saved to disk anyway.
+* During arguments parsing, it should be checked if snapshot argument is true. If it's true one of arguments must be positive. In case both config arguments are invalid (non positive values) an exception should be thrown. In case config arguments are set but snapshot mode is not triggered, both config arguments should be set to invalid values (non positive) and adequate log should be displayed.
+* This project blocks the possibility to split the bag during runtime because disk save happens once per some amount of time and it is already out of the SequentialWriter scope to interfere.
+* For now implementation assumes implementing separate consumer-producer pattern. Idea is to create separate `SnapshotConsumer` and `SnapshotCache` classes (with different names but sharing same interfaces/deriving most functionality from `CacheConsumer` and `MessageCache`). They use similar mechanism as used in bag splitting. Implementation differs mostly in `SnapshotConsumer::close()` and `SnapshotConsumer::~SnapshotConsumer`. Producer-consumer pattern comes into flushing state only after destructor call. Calling bag change/split doesn't prevent `SnapshotCache` from receiving new messages. Additionally buffers swap is only trigerred after `SnapshotConsumer::close()`, which is expected.
+* It should be considered that when buffer doesn't contain any messages to write, rosbag folder should be **deleted** but for now **it is not designed**.
+
+### Pros
+
+* Reuses vast part of `SequentialWriter` class code
+* Accidental `Recorder` node (in snapshot mode) shutdown creates snapshot
+* Producer-consumer patter reusage grants good performance
+
+
+### Cons
+
+* Prevents bag splitting
+* Problematic configuration, pausing and resuming feature (no direct way to change behavior during runtime)
+* Problematic messages counting and rosbag start timestamping
+* (see `TODO` section)
+
+## In-depth structure changes
+
+```diff
```

Comment:
I don't know if I accidentally implied that you should do this diff thing - but please remove the entire diff section. The design review is not the place to put actual code - this will be out of date as soon as we actually implement it. The design should be the place for high level concepts that don't get out of date.

If you would like to open a separate PR with all these changes actually implementing this feature, that would be great! We can link them together in the descriptions as companions of each other.